test/hotspot/jtreg/gc/g1/TestHumongousConcurrentStartUndo.java

```diff
@@ -80,55 +80,74 @@ public static void main(String[] args) throws Exception {
+                (int)(HeapRegionSize * M * 0.75);
+        // Number of objects to allocate to go above IHOP
+        private static final int NumHumongousObjectAllocations =
+                (int)(((HeapSize - YoungSize) * 80 / 100.0) / HeapRegionSize);
+
+
+        private static void allocateHumongous(int num, Queue keeper) {
             for (int i = 1; i <= num; i++) {
                 if (i % 10 == 0) {
                     System.out.println("Allocating humongous object " + i + "/" + num +
-                                       " of size " + objSize + " bytes");
+                                       " of size " + HumongousObjectSize + " bytes");
                 }
-                byte[] e = new byte[objSize];
+                byte[] e = new byte[HumongousObjectSize];
                 if (!keeper.offer(e)) {
                     keeper.remove();
                     keeper.offer(e);
                 }
             }
         }
 
-        public static void main(String [] args) throws Exception {
-            final int M = 1024 * 1024;
-            // Make humongous object size 75% of region size
-            final int humongousObjectSize =
-                (int)(HeapRegionSize * M * 0.75);
+        private static void runConcurrentUndoCycle(ArrayBlockingQueue a) {
+            // Start from an "empty" heap.
+            WB.fullGC();
+            // The queue only holds one element, so only one humongous object
+            // will be reachable and the concurrent operation should be undone.
+            a = new ArrayBlockingQueue(1);
+            allocateHumongous(NumHumongousObjectAllocations, a);
+            Helpers.waitTillCMCFinished(WB, 1);
+        }
 
-            // Number of objects to allocate to go above IHOP
-            final int humongousObjectAllocations =
-                (int)(((HeapSize - YoungSize) * 80 / 100.0) / HeapRegionSize);
+        private static void runConcurrentMarkCycle(ArrayBlockingQueue a) {
+            // Start from an "empty" heap.
+            WB.fullGC();
+            // Try to trigger a concurrent mark cycle. Block concurrent operation
+            // while we are allocating more humongous objects than the IHOP threshold.
+            // After releasing control, trigger the full cycle.
+            try {
+                System.out.println("Acquire CM control");
+                WB.concurrentGCAcquireControl();
+                WB.concurrentGCRunToIdle();
+
+                allocateHumongous(NumHumongousObjectAllocations, a);
+            } finally {
+                System.out.println("Release CM control");
+                WB.concurrentGCReleaseControl();
+            }
+            // At this point we kept NumHumongousObjectAllocations humongous objects live
+            // in "a" which is larger than the IHOP threshold. Another dummy humongous
+            // allocation must trigger a concurrent cycle that is not an Undo Cycle.
+            allocateHumongous(1, new ArrayBlockingQueue(1));
```

Comment:
Since "a" is not referenced beyond this point, I think it is at least theoretically possible that it, and the objects it references, could be treated as dead, allowing an Undo Cycle here too. A following Reference.reachabilityFench(a) could be used to guarantee the desired lifetime.

Doing so also suggests making "a" local rather than an argument, making its size locally obvious.